### PR TITLE
Add support for directional relative permeabilties

### PR DESCRIPTION
**Pull request by hakonhagland**
See https://github.com/OPM/opm-simulators/pull/4048 for more information

### CODE REVIEW

**Comment by hakonhagland** [comment]
jenkins build this please

**Comment by hakonhagland** [comment]
jenkins build this please

**Comment by hakonhagland** [comment]
jenkins build this opm-models=714 opm-material=532 opm-simulators=4048 please

**Comment by hakonhagland** [comment]
jenkins build this opm-models=714 opm-material=532 opm-simulators=4048 please

**Comment by hakonhagland** [comment]
jenkins build this opm-models=714 opm-material=532 opm-simulators=4048 please

File: tests/test_CopyablePtr.cpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    A<B> a1 {};
+    a1.assign(B{1.1,2});
+    BOOST_CHECK_MESSAGE(a1.get_aptr().get()->a == 1.1, "Able to assign new value to pointer");
+    A<B> a2 = a1;
+    BOOST_CHECK_MESSAGE(a2.get_aptr().get()->a == 1.1, "Able to copy construct a new pointer");
+    a1.assign(B{1.3,3});
+    BOOST_CHECK_MESSAGE(a1.get_aptr().get()->a == 1.3, "Able to access new value to pointer");
+    BOOST_CHECK_MESSAGE(a2.get_aptr().get()->b == 2, "Copied value not affected by parent modification");
+    BOOST_CHECK_MESSAGE(a1.get_aptr().get()->getb() == 3, "Member access operator works");
+    BOOST_CHECK_MESSAGE(a1.get_aptr(), "Boolean context operator works");
```
Comment (atgeirr) [inline tests/test_CopyablePtr.cpp:63]: Suggest you also add the converse check at line 54, i.e. that `a1.get_aptr()` is false at that point.

Comment (hakonhagland) [reply tests/test_CopyablePtr.cpp:64]: Good idea! I have updated the PR accordingly.

**Comment by hakonhagland** [comment]
I see this error in the [jenkins build](https://ci.opm-project.org/job/opm-common-PR-builder/5054/console) output:

```
[...]
required from ‘int Opm::start(int, char**, bool) [with TypeTag = Opm::Properties::TTag::ReservoirBlackOilEcfvProblem]’
/var/lib/jenkins/workspace/opm-common-PR-builder/deps/opm-models/tests/reservoir_blackoil_ecfv.cc:56:49:   required from here
/var/lib/jenkins/workspace/opm-common-PR-builder/deps/opm-models/opm/models/blackoil/blackoilintensivequantities.hh:664:50: error: ‘const Problem’ {aka ‘const class Opm::ReservoirProblem<Opm::Properties::TTag::ReservoirBlackOilEcfvProblem>’} has no member named ‘materialLawManager’; did you mean ‘materialLawParams’?
  664 |         const auto& materialLawManager = problem.materialLawManager();
      |                                          ~~~~~~~~^~~~~~~~~~~~~~~~~~
      |                                          materialLawParams
[...]
```

The error message seems to indecate that `problem` (aka `GetPropType<TypeTag, Properties::Problem>` aka `ReservoirBlackOilEcfvProblem`, see https://github.com/OPM/opm-models/blob/dc412e8fb36e9224f2dfbf29e2ddb525c4ee2a0f/tests/reservoir_blackoil_vcfv.cc#L39  ) does not have a method `materialLawManager()`. Any suggestions on how to fix this?

**Comment by totto82** [comment]
I suggest adding a dummy method in multiphasebaseproblem.hh like what is done for the params. See

https://github.com/OPM/opm-models/blob/dc412e8fb36e9224f2dfbf29e2ddb525c4ee2a0f/opm/models/common/multiphasebaseproblem.hh#L239

**Comment by hakonhagland** [comment]
jenkins build this opm-models=714 opm-material=532 opm-simulators=4048 please

**Comment by hakonhagland** [comment]
>I suggest adding a dummy method in multiphasebaseproblem.hh 

@totto82 Good idea, I have added a method materialLawManagerPtr() in  multiphasebaseproblem.hh see https://github.com/OPM/opm-models/pull/714 and overridden that method in eclproblem.hh, see https://github.com/OPM/opm-simulators/pull/4048. Let's see what jenkins says.

**Comment by totto82** [comment]
Cool. Thanks for the work. I will merge it later today if no more comments. I suggest you add a test case to opm-tests so we can add it for regression testing.

**Comment by OPMUSER** [comment]
@hakonhagland I have a few questions on the implementation with respect to the manual:

- This works with both Type 1 and Type 2 relative permeability tables, correct?
- Does this work for both KRNUM* and KRNUM*- options?
- Does  it work with endpoint scaling?
- Does it work with directional endpoint scaling.
- What about the hysteresis option, is that directional as well?

Sorry about the questions, but my head will hurt if I had to go through the code :-)

If you want me to setup a few test cases to test the above for regression testing then let me know.

**Comment by hakonhagland** [comment]
> I suggest you add a test case to opm-tests

@totto82 Yes, see https://github.com/OPM/opm-tests/pull/809

**Comment by totto82** [comment]
> Does this work for both KRNUM* and KRNUM*- options?
> Does it work with endpoint scaling?
> Does it work with directional endpoint scaling.
> What about the hysteresis option, is that directional as well?

This is currently set to not-supported. The PR only change the first to to true. But based on reading the code I think both 2 and 3 should be supported  or nearly supported. But since it is not tested I think they still should give a warning. I will merge this PRs as is, and suggest that we test 2 and 3 afterwards. Can you make some test cases @OPMUSER ?
 
```
               {1,{true, allow_values<std::string> {"HYSTER"}, "SATOPTS(DIRECT): directional relative permeability assignment option not supported - value ignored"}}, // options
               {2,{false, allow_values<std::string> {"HYSTER"}, "SATOPTS(IRREVERS): reversible directional relative permeability option not supported – value ignored"}}, // IRREVERS
               {3,{false, allow_values<std::string> {"HYSTER"}, "SATOPTS(HYSTER): hysteresis directional relative permeability option not supported - value ignored"}}, // HYSTER
               {4,{false, allow_values<std::string> {"HYSTER"}, "SATOPTS(SURFTENS): capillary pressure surface tension pressure dependency option not supported – value ignored"}}, // SURFTENS
```

**Comment by OPMUSER** [comment]
@totto82 Yes I can make a few test cases to verify the options.

**Comment by OPMUSER** [comment]
@totto82 Will start work on setting up a few tests, unless there are other takers :-)

**Pull request merged**